### ISSUE TITLE
Bio.Cluster.clustercentroids cannot  returned the result, Why?

### ISSUE THREAD

**Issue by KKmushroom** [biopython/biopython#2229]
### Setup

I am reporting a problem with Biopython 1.74, Python version 3.7.3, I am running the following program on the spyder as follows:

```python
from sklearn.preprocessing import MinMaxScaler
from Bio.Cluster import kcluster
from Bio.Cluster import clustercentroids
import pandas as pd

df_full = pd.read_csv('../data/FCM_testdata.csv')
columns = list(df_full.columns)
features = columns[:len(columns)-1]
df = df_full[features]
minMax = MinMaxScaler()
dataset=minMax.fit_transform(df)

clusterid1, error, nfound = kcluster(dataset, nclusters=3, dist='e',npass=100)
cdata, cmask = clustercentroids(dataset,clusterid=clusterid1)
```

(*Please copy and run the above in your Python, and copy-and-paste the output*)
![图片](https://user-images.githubusercontent.com/48720081/63638612-33b30680-c6bd-11e9-9d2b-d3eb768019fc.png)
### Expected behaviour

I want to get the return of clustercentroids, the matrix of the cluster center(in the program above is 'cdata')

### Actual behaviour

But the actual behaviour is that the program runs all the time in the spyder and won't return or any results in the Variable explorer. The prompt in the spyder's IPython console is shown above. Why it cannot return results？

### Steps to reproduce
I am not a programmer, just use this library. In view of the above situation, I hope someone can give suggestions, I am eager to use the clustering method of this library to get the cluster center matrix. Thank you for your help!
(*Please fill this in *)

**Comment by mdehoon** [comment]
Can you add some print statements to the script to find out at which point the program hangs? It is not clear now which statement causes the program to hang; the problem may occur before the call to kcluster or clustercentroids.

**Comment by KKmushroom** [comment]
@mdehoon Thank you for your reply. I added a few printed statements to the original manuscript and debugged the program:
`from sklearn.preprocessing import MinMaxScaler
from Bio.Cluster import kcluster
from Bio.Cluster import clustercentroids
import pandas as pd

df_full = pd.read_csv('../data/FCM_testdata.csv')
print(df_full.head(5))
print('size of df_full with real_labels:',df_full.shape)
columns = list(df_full.columns)
features = columns[:len(columns)-1]
df = df_full[features]
print(df.head(5))
print('size of df without real_labels:',df.shape)
minMax = MinMaxScaler()
print(minMax)
dataset=minMax.fit_transform(df)
print(dataset[0:10,:])

clusterid1, error, nfound = kcluster(dataset, nclusters=3, dist='e',npass=100)
cdata, cmask = clustercentroids(dataset,clusterid=clusterid1)`

the print results are shown as follows:


The program hangs the last line: ` cdata, cmask = clustercentroids(dataset,clusterid=clusterid1)`
I debug this function again, adding some print statements to its contents, as follows:
`def clustercentroids(data, mask=None, clusterid=None, method='a',
                     transpose=False):
    data = __check_data(data)
    print('the data type is:',type(data),'data size is:',data.shape)
    mask = __check_mask(mask, data.shape)
    print('the mask value is:',mask[0:5,:])
    print('the input clusterid arg value is:',clusterid[0:5])
    print('the input clusterid type is:',type(clusterid))
    if clusterid is None:
        n = data.shape[1] if transpose else data.shape[0]
        clusterid = numpy.zeros(n, dtype='intc')
    else:
        clusterid = numpy.require(clusterid, dtype='intc', requirements='C')
        print(clusterid)
    return _cluster.clustercentroids(data, mask, clusterid, method, transpose)`

I have printed out the printable output as much as possible, I am not sure if it meets your requirements. If you need me to provide more printing, please let me know where you need to print and view, thank you！
The program is hanging when it reaches the 'return xxx' statement, and Note that the all variables are removed . The results of these print statements output from top to bottom are:
`ipdb> > j:\anaconda3\lib\site-packages\bio\cluster\__init__.py(462)clustercentroids()
    460     """
    461     data = __check_data(data)
--> 462     print('the data type is:',type(data),'data size is:',data.shape)
    463     mask = __check_mask(mask, data.shape)
    464     print('the mask value is:',mask[0:5,:])

ipdb> the data type is: <class 'numpy.ndarray'> data size is: (500, 2)`

`ipdb> > j:\anaconda3\lib\site-packages\bio\cluster\__init__.py(464)clustercentroids()
    462     print('the data type is:',type(data),'data size is:',data.shape)
    463     mask = __check_mask(mask, data.shape)
--> 464     print('the mask value is:',mask[0:5,:])
    465     print('the input clusterid arg value is:',clusterid[0:5])
    466     print('the input clusterid type is:',type(clusterid))

ipdb> the mask value is: [[1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]]`

`> j:\anaconda3\lib\site-packages\bio\cluster\__init__.py(465)clustercentroids()
    463     mask = __check_mask(mask, data.shape)
    464     print('the mask value is:',mask[0:5,:])
--> 465     print('the input clusterid arg value is:',clusterid[0:5])
    466     print('the input clusterid type is:',type(clusterid))
    467     if clusterid is None:

ipdb> the input clusterid arg value is: [0 0 0 0 0]`

`> j:\anaconda3\lib\site-packages\bio\cluster\__init__.py(466)clustercentroids()
    464     print('the mask value is:',mask[0:5,:])
    465     print('the input clusterid arg value is:',clusterid[0:5])
--> 466     print('the input clusterid type is:',type(clusterid))
    467     if clusterid is None:
    468         n = data.shape[1] if transpose else data.shape[0]

ipdb> the input clusterid type is: <class 'numpy.ndarray'>`

`ipdb> > j:\anaconda3\lib\site-packages\bio\cluster\__init__.py(472)clustercentroids()
    470     else:
    471         clusterid = numpy.require(clusterid, dtype='intc', requirements='C')
--> 472         print(clusterid)
    473     return _cluster.clustercentroids(data, mask, clusterid, method, transpose)
    474 

ipdb> [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0
 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0 1 1 0 1 1 1
 1 1 1 0 1 0 1 0 1 0 1 0 0 0 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 0 1 0 1 1 1 1 1
 1 1 0 1 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 2 1
 1 1 1 1 2 2 1 1 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]`
Here the clustried is a matrix, which elements are some labels such as 0,1,2. But what I expected isa matrix, size(n_clusters, n_datafeatures), which elements are cluster centroid. Obviously this is not my hope.

**Comment by mdehoon** [comment]
Can you provide the file `FCM_testdata.csv`? Then I can debug it.

**Comment by KKmushroom** [comment]
Thank you for your reply, here is the test data file, please download it.

I am looking forward to your early reply, thank you very much!


------------------ 原始邮件 ------------------
发件人: "mdehoon"<notifications@github.com>;
发送时间: 2019年8月26日(星期一) 上午8:27
收件人: "biopython/biopython"<biopython@noreply.github.com>;
抄送: "798266099"<798266099@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [biopython/biopython] Bio.Cluster.clustercentroids cannot returned the result, Why? (#2229)




Can you provide the file FCM_testdata.csv? Then I can debug it.

**Comment by KKmushroom** [comment]
@mdehoon The data file used for the test has been sent to you by email. Have you received it?

**Comment by mdehoon** [comment]
@KKmushroom  I have not received it yet.

**Comment by KKmushroom** [comment]
@mdehoon I have uploaded my test program and test data to my repository,  Please go to this address to download: [KKmushroom/biopython-learn
](https://github.com/KKmushroom/biopython-learn)

**Comment by mdehoon** [comment]
@KKmushroom  Thanks. Can you try the updated code in pull request #2231  ?

**Comment by KKmushroom** [comment]
@mdehoon Hello! Please tell me how to use (or download) your updated code? I cloned your biopython project, but I found that the contents of the __init__.py and clustermodule.c files are not like the fix clustercentroids you modified. Do I need to go into the project myself and make changes to the changes you made, or do I need to reinstall Biopython? I don't understand it very much, Please give me some suggestions, thank you!

**Comment by mdehoon** [comment]
@KKmushroom  Are you working on Windows, Mac, or Linux?

**Comment by KKmushroom** [comment]
@mdehoon I working on Windows

**Comment by mdehoon** [comment]
@KKmushroom  Did you recompile the C code in `Bio/Cluster/clustermodule.c`?
Also, can you try the latest version (1.59) of Pycluster available here: 
http://bonsai.hgc.jp/~mdehoon/software/cluster/software.htm#pycluster
You will find precompiled installers there for windows.
This version of Pycluster is identical to the fixed `Bio.Cluster`; just replace "Bio.Cluster" by "Pycluster" in your script.

**Comment by KKmushroom** [comment]
@mdehoon Thank you！

**Comment by mdehoon** [comment]
@KKmushroom  When you use Pycluster, do you still notice any bugs?

**Comment by Jarlan1207** [comment]
@mdehoon Hi, I have the same problem when i working on Linux. Can u tell me what should i do?

**Comment by mdehoon** [comment]
@Jarlan1207  See my comment of August 30.

**Comment by Jarlan1207** [comment]
@mdehoon Sorry, I can't find it. Could u give me a link about it

**Comment by peterjc** [comment]
@Jarlan1207 he meant further up on issue on GitHub, describing how to try the standalone version of the clustering code (which is available separately from Biopython). Link:

https://github.com/biopython/biopython/issues/2229#issuecomment-526595459

**Comment by Jarlan1207** [comment]
@peterjc Thank you

**Comment by mdehoon** [comment]
@KKmushroom  @Jarlan1207  Do you still see this bug with the recently released Biopython 1.75?

**Comment by mdehoon** [comment]
@KKmushroom  @Jarlan1207  If the bug still exists in Biopython 1.75, please let us know before the end of 2019.

**Comment by mdehoon** [comment]
Closing this, since as far as I know this has been fixed.
@KKmushroom please let us know if you still see this bug.
